utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt

```diff
@@ -0,0 +1,158 @@
+import android.text.style.LeadingMarginSpan
+import org.oppia.util.R
+import org.oppia.util.parser.CustomTagHandler.ListItemTag
+import org.oppia.util.parser.StringUtils.LI_TAG
+import org.oppia.util.parser.StringUtils.OL_TAG
+import org.oppia.util.parser.StringUtils.UL_TAG
+import org.xml.sax.XMLReader
+import java.util.*
+
+// TODO(#562): Add screenshot tests to check whether the drawing logic works correctly on all devices.
+/**
+ * Implements support for ordered ({@code <ol>}) and unordered ({@code <ul>}) lists in to Android TextView.
+ * <ul> and <ol> tags are added to the [listParents] Stack and removed when the closing tag is reached.
+ * <li> tags are handled [ListItemTag] instance corresponding to the parent tag.
+ *
+ * Reference: https://github.com/daphliu/android-spannable-list-sample/tree/master/app/src/main/java/com/daphneliu/sample/listspansample
+ */
+class CustomTagHandler(
+  private val context: Context
+) : Html.TagHandler {
+  private val indent = context.resources.getDimensionPixelSize(R.dimen.bullet_leading_margin)
+  private val listItemIndent = indent * 2
+  private val bulletSpan = BulletSpanWithRadius(context, indent)
+
+  private val listParents: Stack<ListItemTag> = Stack<ListItemTag>()
+
+  override fun handleTag(
+    opening: Boolean,
+    tag: String,
+    output: Editable,
+    xmlReader: XMLReader?
+  ) {
+    when (tag) {
+      UL_TAG -> {
+        if (opening) {
+          listParents.push(UnorderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      OL_TAG -> {
+        if (opening) {
+          listParents.push(OrderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
+          throw EmptyStackException()
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
+    /**
+     * Opens a new list item.
+     *
+     * @param text
+     */
+    open fun openItem(text: Editable) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val len = text.length
+      text.setSpan(this, len, len, Spanned.SPAN_MARK_MARK)
+    }
+
+    /**
+     * Closes a list item.
+     *
+     * @param text
+     * @param indentation
+     */
+    fun closeItem(text: Editable, indentation: Int) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val replaces = getReplaces(text, indentation)
+      val len = text.length
+      val listTag = getLast(text)
+      val where = text.getSpanStart(listTag)
+      text.removeSpan(listTag)
+      if (where != len) {
```

Comment:
Please add a comment explaining why this if check is needed (it may not be clear, and isn't actually clear to me at the moment).

Reply:
We need to check if there is span for the text. If not add span.